torch/library.h

```diff
@@ -664,6 +667,24 @@ class TorchLibraryInit final {
   ); \
   void TORCH_LIBRARY_FRAGMENT_init_ ## ns ## _ ## k (torch::Library& m)
 
+/// \private
+///
+/// This macro should only be used in a few legacy areas.
+/// This macro is a version of TORCH_LIBRARY() that:
+///     - doesn't enforce that there is only library for the given namespace
+///     - takes in a unique identifier that it appends to the names
+///       of each static variable that it creates
+/// It effectively lets you define multiple TORCH_LIBRARY_FRAGMENT blocks
+/// for the same namespace within the same translation unit, avoiding naming collisions.
+#define TORCH_LIBRARY_FRAGMENT_UNIQUE(ns, m, uid) \
```

Comment:
I remember suggesting in our 1:1 that you might be able to use `C10_ANONYMOUS_VARIABLE` to avoid the need for `UNIQUE` here. Did you end up trying this / did it not work for some reason? In fact, the old macros seem to make use of `__COUNTER__` directly, which suggests that it is supported on all platforms we care about.

Reply:
The main problem I have with those (`C10_ANONYMOUS_VARIABLE`/`__COUNTER__`) is that the `TORCH_LIBRARY..` macro's need to re-use whatever variable is created- that doesn't fit with the way that `__COUNTER__` works, which is that it's incremented on every use.

It miiiight be possible to fix that with some extra macro logic (like in [this example](https://stackoverflow.com/questions/4719568/reusable-preprocessor-counter), but it would be pretty ugly (I think I'd need to tailor a new "reuse" macro for each existing library macro.

Unless there's an easier way- then I don't think the ugliness there is better than having to change the regex that you mentioned above.

Reply:
The regex values that are used in operator static analysis are [here](https://github.com/pytorch/pytorch/blob/4e6f2440d84a77bd251b7d7e573b0b33e4f3fe5c/tools/code_analyzer/run_analyzer.sh#L23).

I took a look at them, and they shouldn't require any changes to match on the new variable names that I introduced in TORCH_LIBRARY<FRAGMENT|IMPL>_UNIQUE:
```
^.*TORCH_LIBRARY_FRAGMENT_init_([^(]+)(\(.*)?$
^.*TORCH_LIBRARY_IMPL_init_([^(]+)_([^_]+)(\(.*)?$
```
Between the `_init_` and the left paren of the function, both bits of regex match on arbitrary characters (up until the left paren), so adding a counter to the end of the variable name shouldn't affect which ones do/don't get matched.

I wasn't aware of this operator static analysis though- @ljk53 is there anything to watch out for, given that this PR (and the corresponding changes in fbcode) will be updating the caffe2 ops + fbcode ops to use these macros? Those ops will now all become a part of the analysis.

Reply:
> Unless there's an easier way- then I don't think the ugliness there is better than having to change the regex that you mentioned above.

The payoff for having the double macro structure is you get to avoid having both UNIQUE and non-UNIQUE public facing API. The test is: internal complication is not as bad, as long as it is encapsulated, but if the complication leaks out (as it does for UNIQUE), we should try to avoid it.

Reply:
> Between the _init_ and the left paren of the function, both bits of regex match on arbitrary characters (up until the left paren), so adding a counter to the end of the variable name shouldn't affect which ones do/don't get matched.

I don't think this analysis is enough. The regex is not only used to match; the groups matter too (as seen in the use of `return pattern.pattern->sub("\\1", functionName) + "::";`. In particular, if there is extra goop after the namespace, it will be picked up too and you will infer the wrong namespace here.

By the way, this is another good reason to not have two variants of the macro. I think the reason why the existing tests do not catch this is they do not exercise the Caffe2 exported operator codepath (since you're not really supposed to put Caffe2 operators in selective build). If you applied the anonymous variable treatment universally, I bet the selective build tests would have started failing.

Reply:
You're completely right, thanks.

> The payoff for having the double macro structure is you get to avoid having both UNIQUE and non-UNIQUE public facing API

I guess I thought that we could still consider those macros "private" if they were only used within pytorch source code (which is probably wrong). Also, it doesn't actually look that hard to add in the extra macro logic (I'll update the PR shortly with that). Tentatively, I'm updating the existing macros to call out to (truly private) versions of each macro that take in a unique identifier.

> In particular, if there is extra goop after the namespace, it will be picked up too and you will infer the wrong namespace here

Yeah, I didn't look closely enough at this. I just looked at it some more and finally figured out how to run the op analyzer and see / play with the results. With the macro changes that I'm planning on adding, you're right that I think I'll need to update the regex to make sure it correctly pulls out the namespace. As an aside, I think that the existing `TORCH_LIBRARY_FRAGMENT` macro isn't getting its namespace read properly either, since there's a stray `_k` appended to the variable (that gets included in the namespace in the regex) [here](https://github.com/pytorch/pytorch/blob/4e6f2440d84a77bd251b7d7e573b0b33e4f3fe5c/torch/library.h#L659). I'm not sure what/if that actually breaks anything, but it would only affect the ops where that macro is used, [here](https://github.com/pytorch/pytorch/blob/4e6f2440d84a77bd251b7d7e573b0b33e4f3fe5c/aten/src/ATen/native/RNN.cpp#L1862)

Reply:
> I guess I thought that we could still consider those macros "private" if they were only used within pytorch source code (which is probably wrong)

You're right that it's a little wibbly wobbly. One relevant abstraction barrier here is things that are header internal, versus things that get used outside of the header. That's the sense of privacy I'm going for here, though it's definitely a convention.

>  As an aside, I think that the existing TORCH_LIBRARY_FRAGMENT macro isn't getting its namespace read properly either, since there's a stray _k appended to the variable (that gets included in the namespace in the regex) here.

Oops, that sounds like it indeed would be problematic. Actually, the k is probably due to copy paste from `TORCH_LIBRARY_IMPL` and could just be deleted entirely. It's probably not hitting tests because all the registrations in PyTorch proper are just using `TORCH_LIBRARY`; we used to generate fragment registrations for the actual mobile deploy but (1) we never run code analysis on that, and (2) we got rid of that code.